private/bufpkg/bufcheck/bufcheckserver/internal/buflintvalidate/field.go

```diff
@@ -1037,6 +1014,104 @@ func checkExampleValues(
+	fieldDescriptor protoreflect.FieldDescriptor,
+	resolver protodesc.Resolver,
+) (
+	newParentMapFieldDescriptor protoreflect.FieldDescriptor,
+	newFieldDescriptor protoreflect.FieldDescriptor,
+	err error,
+) {
+	// We need to determine the "root" descriptor: if parentMapFieldDescriptor
+	// is set, then fieldDescriptor points to a MapKey or MapValue field, but we
+	// want to create a synthetic clone of the map field itself.
+	rootFieldDescriptor := fieldDescriptor
+	if parentMapFieldDescriptor != nil {
+		rootFieldDescriptor = parentMapFieldDescriptor
+	}
+	originalFile := rootFieldDescriptor.ContainingMessage().ParentFile()
+	newFileDescriptorProto := &descriptorpb.FileDescriptorProto{
+		Name:    proto.String(":synthetic:"),
+		Package: proto.String("__SYNTHETIC__"),
+		Syntax:  proto.String("proto3"),
+		Dependency: []string{
+			originalFile.Path(),
+		},
+	}
+
+	// Ensure that enum and message types are available to be resolved. It's a
+	// bit tricky to get this right, so we just import everything the original
+	// file containing the root message imports (and that file, too.)
+	for i := 0; i < originalFile.Imports().Len(); i++ {
+		newFileDescriptorProto.Dependency = append(
+			newFileDescriptorProto.Dependency,
+			originalFile.Imports().Get(i).Path(),
+		)
+	}
+
+	// Create a clone of the root field descriptor. Since our target message is
+	// going to be empty aside from this field, we also clear the OneofIndex.
+	newFieldDescriptorProto := protodesc.ToFieldDescriptorProto(rootFieldDescriptor)
+	newFieldDescriptorProto.OneofIndex = nil
+
+	newMessageDescriptorProto := &descriptorpb.DescriptorProto{
+		Name:  proto.String(string(rootFieldDescriptor.ContainingMessage().Name())),
+		Field: []*descriptorpb.FieldDescriptorProto{newFieldDescriptorProto},
+	}
+
+	// The map field and entry message have to be declared in the same scope, so
+	// we need to make a clone of the map entry type itself. This needs to be
+	// nested under the synthetic message.
+	if parentMapFieldDescriptor != nil {
+		// Shouldn't happen, but we don't want to panic if it does.
+		if !parentMapFieldDescriptor.IsMap() {
+			return nil, nil, errors.New("parentMapFieldDescriptor is not a map")
+		}
+		newMapEntryMessageDescriptor := &descriptorpb.DescriptorProto{
+			Name: proto.String(string(parentMapFieldDescriptor.Message().Name())),
+			Field: []*descriptorpb.FieldDescriptorProto{
+				protodesc.ToFieldDescriptorProto(parentMapFieldDescriptor.MapKey()),
+				protodesc.ToFieldDescriptorProto(parentMapFieldDescriptor.MapValue()),
+			},
+			Options: &descriptorpb.MessageOptions{
+				MapEntry: proto.Bool(true),
+			},
+		}
+		newMessageDescriptorProto.NestedType = append(
+			newFileDescriptorProto.MessageType,
+			newMapEntryMessageDescriptor,
+		)
+		newFieldDescriptorProto.Type = descriptorpb.FieldDescriptorProto_TYPE_MESSAGE.Enum()
+		newFieldDescriptorProto.TypeName = proto.String(newMapEntryMessageDescriptor.GetName())
```

Comment:
This isn't necessary. If it was originally a map field, these are already set this way, from the call to `protodesc.ToFieldDescriptorProto`.

Reply:
Oof, never mind. That is only true of the first one, the type enum. This is correct that it must re-write the type name.